File: design_patterns/creational/builder/builder.py
from abc import ABC, abstractstaticmethod

class RTFReader:
    """ Director """
    __builder = None

    def SetBuilder(self, builder):
       self.__builder = builder

    def ParseRTF(self, body):
        for token in body:
            if token == 'CHAR':
               self.__builder.ConvertCharacter(token)
            if token == 'FONT':
               self.__builder.ConvertFontChange(token)
            if token == 'PARA':
               self.__builder.ConvertParagraph()


class TextConverter(ABC):
  """ Builder Interface"""

  @abstractstaticmethod
  def ConvertCharacter(self, char):
      pass

  @abstractstaticmethod
  def ConvertFontChange(self, Font):
      pass

  @abstractstaticmethod
  def ConvertParagraph(self):
      pass

class ASCIIConverter(TextConverter):
    """
    ASCII Concrete Builder,
    ignores request to convert anything except plain text.
    """

    def ConvertCharacter(self, char):
        print("ASCII converted character")

    def ConvertFontChange(self, Font):
        print("ASCII does not support conversion for Font")

    def ConvertParagraph(self):
        print("ASCII does not support conversion for paragraphs")

    def GetASCIIText(self):
        print("ASCII representation")

class TeXConverter(TextConverter):
    """
    TeX Concrete Builder
    """

    def ConvertCharacter(self, char):
        print("TeX converted character")

    def ConvertFontChange(self, Font):
        print("TeX converted Font")

    def ConvertParagraph(self):
        print("TeX converted paragraph")

    def GetTeXText(self):
        print("TeX representation")

class TextWidgetConverter(TextConverter):
    """
    Text Widget Concrete Builder
    """

    def ConvertCharacter(self, char):
        print("Text Widget converted character")

    def ConvertFontChange(self, Font):
        print("Text Widget converted Font")

    def ConvertParagraph(self):
        print("Text Widget converted Paragraph")

    def GetTextWidget(self):
        print("Text Widget representation")


def clientCode():
    """
    Client code executed as example
    """
    demoTokens = ['CHAR', 'FONT', 'PARA']

    rtfReader = RTFReader()
    asciiConverter = ASCIIConverter()
    texConverter = TeXConverter()
    textWidgetConverter = TextWidgetConverter()

    rtfReader.SetBuilder(asciiConverter)
    asciiText = rtfReader.ParseRTF(demoTokens)

    rtfReader.SetBuilder(texConverter)
    texText = rtfReader.ParseRTF(demoTokens)

    rtfReader.SetBuilder(textWidgetConverter)
    textWidgetText = rtfReader.ParseRTF(demoTokens)

if __name__ == "__main__":
    clientCode()

File: container2/design_patterns/creational/builder/builder.js
/**
 * Builder Director
 */
class RTFReader {
    _builder = undefined

    setBuilder = function(builder) {
        this._builder = builder
    }

    parseRTF = function(body) {
        for (const token of body) {
            if (token === "CHAR") this._builder.ConvertCharacter(token)
            if (token === "FONT") this._builder.ConvertFont(token)
            if (token === "PARA") this._builder.ConvertParagraph()
        }
    }
}

/**
 * Builder Interface
 */
class TextConverter {
    ConvertCharacter = function() {}
    convertFont = function() {}
    ConvertParagraph = function() {}
}

/**
 * ASCII Concrete Builder
 */
class ASCIIConverter {
    ConvertCharacter = function() { console.log("ASCII Character") }
    ConvertFont = function() { console.log("ASCII Font") }
    ConvertParagraph = function() { console.log("ASCII Paragraph") }
}

/**
 * TeX Concrete Builder
 */
class TeXConverter {
    ConvertCharacter = function() { console.log("TeX Character") }
    ConvertFont = function() { console.log("TeX Font") }
    ConvertParagraph = function() { console.log("TeX Paragraph") }
}

/**
 * Example Code
 */
function demoClient() {
    demoTokens = ["CHAR", "FONT", "PARA"]

    rtfReader = new RTFReader() // Instance a Director
    asciiConverter = new ASCIIConverter() // Instance a ConcreteBuilder
    texConverter = new TeXConverter() // Instanca a difference ConcreteBuilder

    rtfReader.setBuilder(asciiConverter)
    asciiText = rtfReader.parseRTF(demoTokens)

    rtfReader.setBuilder(texConverter)
    texText = rtfReader.parseRTF(demoTokens)
}

demoClient()
